### PR TITLE
Cross-platform media controller for video playback

### LINKED ISSUE #1811: VideoPlayer controls buttons not showing
Hi guys,
With cocos2d-x when you start the video VideoPlayer, and tap on the video playing, the video will stop and you will have the video controls showing. This is the normal behavior for any app (not only for coco2d-x apps):

With Axmol, when you tap on the video playing, the behavior is completely different (it doesn’t make it user friendly):
1) When you tap on the video playing, the video doesn’t stop playing. It should stop playing.

2) The video controls will not appear. They should appear.

Please see attached screenshots to see how it should look. And what is happening with Axmol.

This is a normal case (the video controls are accessible). and when you tap on it, it will stop playing:

normalBehavior

This is Axmol bug, where the video controls are not accessible an when you tap on it it will not stop playing:

Screenshot 2024-04-10 at 6 20 39 PM

You can try with the url provided in the VideoPlayer test:  https://storage.googleapis.com/exoplayer-test-media-1/mp4/android-screens-25s.mp4

Thank you,
Stay safe

- axmol version: 2.1.2
- devices test on: iPhon 11, 12, 13, 14
- developing environments
   - NDK version: N/A
   - Xcode version: N/A
   - Visual Studio: 
      - VS version: N/A
      - MSVC version: N/A
      - Windows SDK version: N/A
   - cmake version:  3.29.0
Steps to Reproduce:

1. Build a small test App with the video VideoPlayer (or use the test App provided by Axmol)
2. use the url provided by Axmol to display the video (https://storage.googleapis.com/exoplayer-test-media-1/mp4/android-screens-25s.mp4)
3. When the video is playing, tap on it

Thank you,
Stay safe

### PR DESCRIPTION
**Pull request by rh101**
## Describe your changes

This PR is to add a media controller to video playback, which works across all supported platforms and media engine backends.

- [x] AndroidMediaEngine
- [x] AvfMediaEngine
- [x] MfMediaEngine
- [x] VlcMediaEngine
- [x] WmfMediaEngine 

@halx99 Is the VLC engine only for Linux? Attempting to compile it on Windows ended up with this error:
`3>Error copying directory from "G:/game_engines/axmol/3rdparty/vlc/win/lib/vlc/plugins" to "G:/dev/public/tests/videotest/build_x64/bin/videotest/Debug/plugins".`

I don't see the `vlc` folder in the `axmol/3rdparty` either, and I'm not sure what else I need to do besides enable it via `set(AX_ENABLE_VLC_MEDIA ON CACHE BOOL "" FORCE)`.

By default the player controls are off.  Would it be preferrable to have them disabled by default?

At the moment, to enable or disable them, just use `VideoPlayer::setUserInputEnabled()`. For example:
```
auto* videoPlayer = ui::VideoPlayer::create();
videoPlayer->play();
videoPlayer->setUserInputEnabled(true);
```

The default media controller can also be overriden with a custom version, and applied to the video player via `MediaPlayer::setMediaController(MediaController* controller)`.

The Android implementation required a work-around, since there was no way to retrieve the content duration and current time directly.  It has to happen in callbacks, but I'm not sure if the solution implemented in this PR is the most appropriate, so please review that, specifically the code in `AxmolMediaEngine.java`, method `processVideoFrame`, where it calls back with the current video playback time.

## Issue ticket number and link
#1811 

### For each PR
- [ ] Add Copyright if it missed:   
      - `"Copyright (c) 2019-present Axmol Engine contributors (see AUTHORS.md)."`
- [x] I have performed a self-review of my code.
       
   Optional:
   - [x] I have checked readme and add important infos to this PR.
   - [ ] I have added/adapted some tests too.
          
### For core/new feature PR
- [x] I have checked readme and add important infos to this PR.
- [ ] I have added thorough tests.

### CODE REVIEW

**Comment by aismann** [comment]
@rh101 
Please add a copyright header/comment on each file too

**Comment by halx99** [comment]
> @halx99 Is the VLC engine only for Linux? Attempting to compile it on Windows ended up with this error:
> `3>Error copying directory from "G:/game_engines/axmol/3rdparty/vlc/win/lib/vlc/plugins" to "G:/dev/public/tests/videotest/build_x64/bin/videotest/Debug/plugin`

Yes, it's only for linux, and axmol prebuilt libvlc not work for current axmol version

**Comment by rh101** [comment]
Currently in the process of testing the Linux version, and fixing a few issues related to content scaling.  I'm trying to convert this PR to a draft with the button on github, but it's not letting me, so for the time being, please don't merge anything in.

**Comment by rh101** [comment]
Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system. 

I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.

Any suggestions?

**Comment by halx99** [comment]
> Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system.
> 
> I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.
> 
> Any suggestions?

Use drawNode to draw control buttons shape?

**Comment by halx99** [comment]
> By default the player controls are on. Would it be preferrable to have them disabled by default?

disable by default maybe better

**Comment by halx99** [comment]
Maybe the `UIMediaPlayer::setUserInputEnabled` can be removed, instead use new API: setMediaControllerEnabled ?

**Comment by halx99** [comment]
> > Linux version runs, but I'm having issues with missing media player control font characters. The Android version is also missing the play button font character, yet all the other buttons appear correctly, so another issue with the fonts on that system.
> > I'm not sure how to resolve this without using image files, but with image files, to not look pixelated, they end up as 3-4K file each, so the resulting base64 encoded PNGs are pretty large.
> > Any suggestions?
> 
> Use drawNode to draw control buttons shape?

Or embed textures due to the buttons shape is simply. for example progress bar can use one 2x2 white texture, and embed arrow, stop textures

File: core/ui/UIMediaPlayer.cpp [hunk 1]
```diff
@@ -136,20 +138,659 @@ struct PrivateVideoDescriptor
         PS_SET_UNIFORM_R(ps, "colorTransform", colorTransform);
     }
 };
+
+std::unique_ptr<MediaEngineFactory> _meFactory = MediaEngineFactory::create();
+
+const char* CIRCLE_IMAGE =
+    "iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/"
```
Comment (halx99) [inline core/ui/UIMediaPlayer.cpp]: convert to binary maybe better

Comment (rh101) [reply core/ui/UIMediaPlayer.cpp]: Oh! I actually missed that. We don't need it anymore, since I can just use the DrawNode code to generate it too.

**Review by halx99 — APPROVED** [review]

**Comment by halx99** [comment]
Hi @rh101
@asnagni said can't see controller buttons on iOS, could you take a look? 
you can login to discord to view detail conversations

**Comment by rh101** [comment]
> Hi @rh101 @asnagni said can't see controller buttons on iOS, could you take a look? you can login to discord to view detail conversations

I'll load it up right now and check

**Pull request merged**